### DISCUSSION TITLE
A .js file missing/renamed in matplotlib?

### DISCUSSION THREAD

**Discussion by savakarrohan** [jupyterlite/jupyterlite#983] [Q&A] [answered]
Let me start off by saying that a typical import of matplotlib works fine. I am finding an issue in a slightly different problem.

## My Package
I have built and am serving the following package [https://github.com/savakarrohan/backtrader](https://github.com/savakarrohan/backtrader) as a wheel. I import this package through the `piplite.install` method with an appropriate local development link ending with `.whl` as mentioned in documentation.

The imports don't appear to have a problem. as functions and classes are working well from the package. 

## The issue
A separate part of my code involves plotting using the `matplotlib.pyplot`. There are errors thrown whenever I import my package. 
```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[16], line 1
----> 1 mpf.plot(stockPandasTot["LT.NS"].iloc[::-1],type='candle',mav=(5,15),volume=True,style='charles')

File /lib/python3.10/site-packages/mplfinance/plotting.py:951, in plot(data, **kwargs)
    949         plt.close(fig)
    950 elif not config['returnfig']:
--> 951     plt.show(block=config['block']) # https://stackoverflow.com/a/13361748/1639359
    952     if config['closefig'] == True or (config['block'] and config['closefig']):
    953         plt.close(fig)

File /lib/python3.10/site-packages/matplotlib/pyplot.py:389, in show(*args, **kwargs)
    345 """
    346 Display all open figures.
    347 
   (...)
    386 explicitly there.
    387 """
    388 _warn_if_gui_out_of_main_thread()
--> 389 return _get_backend_mod().show(*args, **kwargs)

File /lib/python3.10/site-packages/matplotlib/backends/backend_nbagg.py:258, in _BackendNbAgg.show(block)
    255 interactive = is_interactive()
    257 for manager in managers:
--> 258     manager.show()
    260     # plt.figure adds an event which makes the figure in focus the
    261     # active one. Disable this behaviour, as it results in
    262     # figures being put as the active figure after they have been
    263     # shown, even in non-interactive mode.
    264     if hasattr(manager, '_cidgcf'):

File /lib/python3.10/site-packages/matplotlib/backends/backend_nbagg.py:89, in FigureManagerNbAgg.show(self)
     87 def show(self):
     88     if not self._shown:
---> 89         self.display_js()
     90         self._create_comm()
     91     else:

File /lib/python3.10/site-packages/matplotlib/backends/backend_nbagg.py:85, in FigureManagerNbAgg.display_js(self)
     81 def display_js(self):
     82     # XXX How to do this just once? It has to deal with multiple
     83     # browser instances using the same kernel (require.js - but the
     84     # file isn't static?).
---> 85     display(Javascript(FigureManagerNbAgg.get_javascript()))

File /lib/python3.10/site-packages/matplotlib/backends/backend_nbagg.py:113, in FigureManagerNbAgg.get_javascript(cls, stream)
    111 else:
    112     output = stream
--> 113 super().get_javascript(stream=output)
    114 output.write((pathlib.Path(__file__).parent
    115               / "web_backend/js/nbagg_mpl.js")
    116              .read_text(encoding="utf-8"))
    117 if stream is None:

File /lib/python3.10/site-packages/matplotlib/backends/backend_webagg_core.py:508, in FigureManagerWebAgg.get_javascript(cls, stream)
    505 else:
    506     output = stream
--> 508 output.write((Path(__file__).parent / "web_backend/js/mpl.js")
    509              .read_text(encoding="utf-8"))
    511 toolitems = []
    512 for name, tooltip, image, method in cls.ToolbarCls.toolitems:

File /lib/python3.10/pathlib.py:1132, in Path.read_text(self, encoding, errors)
   1128 """
   1129 Open the file in text mode, read it, and close the file.
   1130 """
   1131 encoding = io.text_encoding(encoding)
-> 1132 with self.open(mode='r', encoding=encoding, errors=errors) as f:
   1133     return f.read()

File /lib/python3.10/pathlib.py:1117, in Path.open(self, mode, buffering, encoding, errors, newline)
   1115 if "b" not in mode:
   1116     encoding = io.text_encoding(encoding)
-> 1117 return self._accessor.open(self, mode, buffering, encoding, errors,
   1118                            newline)

FileNotFoundError: [Errno 44] No such file or directory: '/lib/python3.10/site-packages/matplotlib/backends/web_backend/js/mpl.js'
```
That error is only thrown when I import my particular package, note the command used here `mpf.plot()` is not from my package. 

## My attempts
I have tried to change the backends being used by matplotlib by using `matplotlib.use("Tkagg")` etc but I understand those are not available, so I have to use `Agg`.
I have tried combinations of ordering the imports, because I thought it could be a referencing error.
I have gone through the metadata for the packages on pyodide.

Does anyone have any leads on how I can move forward in this problem?

**Comment by savakarrohan** [comment]
Is it also possible to see the location of where the package files are imported to while using the application? 
I mean I would like to view where the packages are stored and see the file system there when installed using `piplite.install`

**Comment by bollwyvl** [comment, marked as answer]
> the following package

Does the package actually declare its dependency on `matplotlib`? 

What about  `import matplotlib` before installing/importing this package?

> see the location of where the package files are imported

It's printed out right the log, and can be listed by normal pythonic means, but not available in the UI (#982):

```py
from pathlib import Path
sorted(Path('/lib/python3.10/site-packages/matplotlib').rglob('*'))
```

> I have to use Agg.

If "have to," then that notebook is going to have a bad day. The only backends known/tested to work are:

- the default when using `import matplotlib` (this is provided by pyodide)
  - https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html?path=pyolite/matplotlib.ipynb
- `ipympl`
  - https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html?path=pyolite/ipympl.ipynb
  - this also requires have that as a federated extension, either by having it installed at build time or listing it (and its dependencies) explicitly

Pretty much anything else is not tested, and very probably not going to work at all.

**Reply by savakarrohan** [reply]
> Does the package actually declare its dependency on matplotlib?

Yes, I have included it under setup.py `install_requires=['matplotlib>=3.5.0']`

> What about `import matplotlib` before installing/importing this package?

I did try out various orders of this and it initially appeared that it wasn't working. 
It appears that I had a plotting command from my `backtrader` package prior to the typical `plt` command. 
This `backtrader plotting` is forcing a matplotlib to use `nbagg` which is not supported. 
I still have to debug it but somewhere in one of my plot commands, I am using a different backend for matplotlib and it's causing an issue. 

I believe the default matplotlib backend module is `module://matplotlib_inline.backend_inline`
while the one with ipympl is `module://ipympl.backend_nbagg`
extracted from the examples you have mentioned above using function `matplotlib.get_backend()`

Guess I will get back to this when I debug more on my package and attempt to use these backends somehow. Any tips will be welcome.

**Reply by savakarrohan** [reply]
Hi, I solved it, it was a problem with my package. Thank you!!
